Javadoc improvement org.freeplane.plugin.script.FreeplaneScriptBaseClass

HaaiHenkie opened the pull request:
Fixed broken links in main description
In main description and ConfigProperties description: added reference to importing GroovyStaticImports in case of compiling source code outside Freeplane, see also https://github.com/freeplane/freeplane/issues/2499

HaaiHenkie committed:
Javadoc improvement org.freeplane.plugin.script.FreeplaneScriptBaseClass

Fixed links in main description
Extended description nested class ConfigProperties for use in utility
scripts and addon classes, see also
https://github.com/freeplane/freeplane/issues/2499

dpolivaev commented on freeplane_plugin_script/src/main/java/org/freeplane/plugin/script/FreeplaneScriptBaseClass.java:
```diff
@@ -45,22 +45,30 @@
-	 * Accessor for Freeplane's configuration: In scripts available
-	 * as "global variable" <code>config</code>.
+	 * <p>Accessor for Freeplane's configuration: In scripts available as "global variable" {@code config}</p>
+     *
```
I do not understand what you mean by "utility scripts and addon classes".
If I understand it correctly, the variable is available in all classes compiled by Freeplane, but it is not available in classes compiled outside. Am I right?

HaaiHenkie replied:
I mean that the global variable `config` is only available in scripts. This variable (like `node` and `c`) is **not** available in classes compiled by Freeplane, i.e. [utility scripts](https://docs.freeplane.org/scripting/Your_own_utility_script_library.html) and classes compiled to jar when packaging an add-on.

dpolivaev replied:
I have just commented at https://github.com/freeplane/freeplane/issues/2499#issuecomment-2972700333

HaaiHenkie replied:
So you were right, it is not available when compiling groovy classes outside Freeplane. I adjusted the javadoc accordingly.

dpolivaev replied:
I added another comment here: https://github.com/freeplane/freeplane/issues/2499#issuecomment-2973635300.
Could you check it and see if it makes sense to include it in the documentation?

And I want to highlight something important: it was your commitment — shown in this PR — that made me dive into the code and try to understand what’s really going on.
Your contribution already made a difference, and I truly hope you’ll keep going.

HaaiHenkie replied:
In the main description of `FreeplaneScriptBaseClass` and in the `ConfigProperties` description, I have added a reference to importing `GroovyStaticImports` in case of compiling source code outside Freeplane.

I looked into `GroovyStaticImports`. I want to make sure if I understand it well and I would like to make some suggestions in the end.

It seems that `GroovyStaticImports` is the class that is injecting the DSL shortcuts for the 6 global classes available in the Freeplane scripting environment into any utility script or groovy source code on the script classpath. Since it is dated 30 January 2024, it means that before that time you needed to add this imports manually to your groovy source code. That is what I had in mind writing the first PR.

So `GroovyStaticImports` imports the same 6 global classes as `FreeplaneScriptBaseClass`.

In addition `GroovyStaticImports` imports `ScriptUtils` as `scriptUtils` and `ScriptUtils.ignoreCycles()` as `ignoreCycles()`. As far as I know this is not documented so I doubt if anybody is using this imports in utility scripts.

`GroovyStaticImports` does not import the static methods that `FreeplaneScriptBaseClass` provides without prefix, like `N(String id)`, `T(String id)`, `V(String id)` and `ifNull(Object value, Object valueIfNull)`. I don't mind since a hardly ever use these methods. But if someone is using this in a utility script it will give an error.

My first recommendation is to document `GroovyStaticImports`.

My second recommendation is to add `GroovyStaticImports` to the Freeplane API and to remove it as an automatic injection. This has the following advantages:
1. It can be documented as part of the API
2. Globals `scriptUtils` and `ignoreCycles()` are recognised in the IDE
3. If the source code is a groovy class, all other globals are recognised in the IDE (since in this situation their definition is not provided by `freeplane.dsld` or `freeplane.gdsl`)
4. There is no difference between compiling utility scripts or source code inside or outside Freeplane

HaaiHenkie committed:
Javadoc improvement org.freeplane.plugin.script.FreeplaneScriptBaseClass

Fixed broken links in main description
Extended description nested class ConfigProperties for compiling classes
outside Freeplane, see also
https://github.com/freeplane/freeplane/issues/2499

HaaiHenkie committed:
Javadoc improvement org.freeplane.plugin.script.FreeplaneScriptBaseClass

Fixed broken links in main description
In main description and ConfigProperties description: added reference to
importing GroovyStaticImports in case of compiling source code outside
Freeplane, see also
https://github.com/freeplane/freeplane/issues/2499
